src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_formula_generator.py

```diff
@@ -140,3 +140,77 @@ def generate(
+
+    def _all_has_category(
+        self, components: set[Component], category: ComponentCategory
+    ) -> bool:
+        return all(component.category == category for component in components)
+
+    def _get_metric_fallback_components(
+        self, components: set[Component]
+    ) -> dict[Component, set[Component]]:
+        """Get primary and fallback components within a given set of components.
+
+        When a meter is positioned before one or more components of the same type (e.g., inverters),
+        it is considered the primary component, and the components that follow are treated
+        as fallback components.
+        If the non-meter component has no meter in front of him, then it is the primary component
+        and has no fallbacks.
+
+        The method iterates through the provided components and assesses their roles as primary
+        or fallback components.
+        If component:
+         * can act as a primary component (e.g., a meter), then it finds its
+        fallbacks components and pairs them together.
+         * can act as a fallback (e.g., an inverter or EV charger), then it finds
+        the primary component for him (usually a meter), and pairs them together.
+         * has no fallback (e.g., an inverter that has no meter attached), then it
+        returns an empty set for that component. This means that the component
+        is a primary component and has no fallbacks.
+
+        Args:
+            components: The component to be analyzed.
+
+        Returns:
+            A dictionary where:
+                * The keys are primary components.
+                * The values are sets of fallback components.
+        """
+        graph = connection_manager.get().component_graph
+
+        fallbacks: dict[Component, set[Component]] = {}
+        for component in components:
+
+            if component.category in (
+                ComponentCategory.INVERTER,
+                ComponentCategory.EV_CHARGER,
+                ComponentCategory.CHP,
+            ):
+                predecessors = graph.predecessors(component.component_id)
+
+                if len(predecessors) == 1:
+                    predecessor = predecessors.pop()
+
+                    if predecessor.category == ComponentCategory.METER:
+                        # Meter is primary component and the other one is one of the
+                        # fallbacks components.
+                        fallbacks.setdefault(predecessor, set()).add(component)
+                        continue
+
+                # This component is primary component with no fallbacks.
+                fallbacks[component] = set()
+            elif component.category == ComponentCategory.METER:
+                successors = graph.successors(component.component_id)
+
+                if (
+                    all(c.category == ComponentCategory.CHP for c in successors)
+                    or all(c.category == ComponentCategory.INVERTER for c in successors)
```

Comment:
Here you'd need to check if they're all the same type as well.  Either all battery inverters or all PV inverters.

Reply:
Right, thanks! The first case is also wrong. I am fixing it now.

Reply:
Done